src/mod_last_rdbms.erl

```diff
@@ -25,52 +25,66 @@
          remove_user/2]).
 
 -spec init(jid:server(), list()) -> ok.
-init(_Host, _Opts) ->
+init(Host, _Opts) ->
+    prepare_queries(Host),
     ok.
 
+%% Prepared query functions
+prepare_queries(Host) ->
+    mongoose_rdbms:prepare(last_select, last, [username],
+                           <<"SELECT seconds, state FROM last WHERE username=?">>),
+    mongoose_rdbms:prepare(last_count_active, last, [seconds],
+                           <<"SELECT COUNT(*) FROM last WHERE seconds > ?">>),
+    mongoose_rdbms:prepare(last_delete, last, [username],
+                           <<"delete from last where username=?">>),
+    rdbms_queries:prepare_upsert(Host, last_upsert, last,
+                                 [<<"username">>, <<"seconds">>, <<"state">>],
+                                 [<<"seconds">>, <<"state">>],
+                                 [<<"username">>]).
+
+execute_get_last(LServer, LUser) ->
+    mongoose_rdbms:execute_successfully(LServer, last_select, [LUser]).
+
+execute_count_active_users(LServer, Seconds) ->
+    mongoose_rdbms:execute_successfully(LServer, last_count_active, [Seconds]).
+
+execute_remove_user(LServer, LUser) ->
+    mongoose_rdbms:execute_successfully(LServer, last_delete, [LUser]).
+
+execute_upsert_last(Host, LUser, Seconds, State) ->
+    InsertParams = [LUser, Seconds, State],
+    UpdateParams = [Seconds, State],
+    UniqueKeyValues = [LUser],
+    rdbms_queries:execute_upsert(Host, last_upsert, InsertParams, UpdateParams, UniqueKeyValues).
+
+%% API functions
 -spec get_last(jid:luser(), jid:lserver()) ->
-    {ok, non_neg_integer(), binary()} | {error, term()} | not_found.
+    {ok, non_neg_integer(), binary()} | not_found.
 get_last(LUser, LServer) ->
-    Username = mongoose_rdbms:escape_string(LUser),
-    case catch rdbms_queries:get_last(LServer, Username) of
-        {selected, []} ->
-            not_found;
-        {selected, [{STimeStamp, Status}]} ->
-            case catch mongoose_rdbms:result_to_integer(STimeStamp) of
-                TimeStamp when is_integer(TimeStamp) ->
-                    {ok, TimeStamp, Status};
-                Reason ->
-                    {error, {invalid_timestamp, Reason}}
-            end;
-        Reason -> {error, {invalid_result, Reason}}
-    end.
+    Result = execute_get_last(LServer, LUser),
+    decode_last_result(Result).
 
 -spec count_active_users(jid:lserver(), non_neg_integer()) -> non_neg_integer().
-count_active_users(LServer, TimeStamp) ->
-    TimeStampBin = integer_to_binary(TimeStamp),
-    WhereClause = <<"where seconds > ", TimeStampBin/binary >>,
-    case rdbms_queries:count_records_where(LServer, <<"last">>, WhereClause) of
-        {selected, [{Count}]} ->
-            mongoose_rdbms:result_to_integer(Count);
-        _ ->
-            0
```

Comment:
This case was removed. Did you make sure it does not happen during normal operation? Is this again the code being overly defensive?

Reply:
I prefer to use `execute_successfully` everywhere. It would report to logs and throw an error.
So, less code here (no need to match on `aborted` and `error`).

Using 0 here is just an attempt to handle the error by faking data.
I am trying to remove this kinda code from backend modules code.

I think the caller (i.e. the main module) should handle the error explicitly (or just fail, it's an option too).